clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java

```diff
@@ -3146,6 +3149,97 @@ public void testInvalidTxnStateIsAnAbortableError() throws Exception {
 
         txnManager.beginTransaction();
     }
+    @Test
+    public void testProducerBatchRetriesWhenPartitionLeaderChanges() throws Exception {
+        Metrics m = new Metrics();
+        SenderMetricsRegistry senderMetrics = new SenderMetricsRegistry(m);
+        try {
+            // SETUP
+            String metricGrpName = "producer-metrics-test-stats-1";
+            long totalSize = 1024 * 1024;
+            BufferPool pool = new BufferPool(totalSize, batchSize, metrics, time,
+                metricGrpName);
+            long retryBackoffMaxMs = 100L;
```

Comment:
How difficult would it be to mock the backoff using a ticker? Too many times having to rely on timing in tests makes them become flaky. This seems like it would low probability here but we should keep it in mind

Reply:
Fair point. The `SenderTest` already mocks `time` using `MockTime`, and that is manually advanced. Is that what u had in mind?

Reply:
That's good to know. My concern came from `time.sleep(2 * retryBackoffMaxMs);` down on line 3223. I'm not sure that is avoidable though

Reply:
`time.sleep(2 * retryBackoffMaxMs);` that is how i ensure that tests waits more than backoff period before retrying.